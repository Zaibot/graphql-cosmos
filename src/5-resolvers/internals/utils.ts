import { FeedResponse } from '@azure/cosmos'
import { GraphQLCosmosPageOutput } from '../../4-resolver-builder/3-typedefs-transformer'
import { Lazy } from '../../typescript'
import { SourceDescriptor } from '../x-descriptors'

export const emptyPageResponse: GraphQLCosmosPageOutput = {
  cursor: null,
  nextCursor: null,
  page: [],
  total: 0,
}

export function graphqlCosmosPageResponse(
  typename: string,
  database: string,
  container: string,
  cursor: string | null,
  query: Lazy<Promise<FeedResponse<{ id: string }>>>,
  count: Lazy<Promise<FeedResponse<number>>>
): GraphQLCosmosPageOutput {
  return {
    get cursor() {
      return Promise.resolve(cursor)
    },
    get nextCursor() {
      return query().then((x) => x.continuationToken)
    },
    get page() {
      return query().then((x) => x.resources.map(wrapSingleSourceDescriptor(typename, database, container)))
    },
    get total() {
      return count().then((x) => x.resources[0])
    },
  }
}

export function wrapSingleSourceDescriptor(typename: string, database: string, container: string) {
  // TODO, use refSingle of datasource
  return (item: { id: string }) =>
    SourceDescriptor.withDescriptor(item, {
      kind: `Single`,
      database,
      container,
      typename,
      id: item.id,
    })
}
