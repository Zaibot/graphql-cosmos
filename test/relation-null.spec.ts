import gql from 'graphql-tag'
import { printSchemaWithDirectives } from 'graphql-tools'
import { createUnitTestContext } from './utils'

describe(`Nulled relations`, () => {
  const dummyTypeDefs = gql`
    type Query {
      dummies: [Dummy] @cosmos
    }

    type Dummy @cosmos(database: "Test", container: "Dummies") {
      id: ID!
      prop: String
      embedded: [Embedded]
      related: Related @cosmos(ours: "relatedId")
    }

    type Embedded {
      prop: String
      related: Related @cosmos(ours: "relatedId")
    }

    type Related @cosmos(database: "Test", container: "Related", ours: "relatedId") {
      id: ID!
      prop: String
      related: Related @cosmos(ours: "relatedId")
    }
  `

  const responses = {
    Dummies: {
      'SELECT c.id FROM c ORDER BY c.id': [{ id: `1` }, { id: `2` }, { id: `3` }, { id: `4` }, { id: `5` }],
      'SELECT c.id, c.prop, c.embedded, c.relatedId FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=1,2,3,4,5)':
        [
          { id: `1`, prop: `text`, relatedId: null, embedded: [{ prop: `text`, relatedId: null }] },
          { id: `2`, prop: `text`, relatedId: undefined, embedded: [{ prop: `text`, relatedId: undefined }] },
          { id: `3`, prop: `text`, relatedId: undefined, embedded: [{ prop: `text` }] },
          { id: `4`, prop: `text`, relatedId: undefined, embedded: [{ prop: `text`, relatedId: `5` }] },
          { id: `5`, prop: `text`, embedded: [{ prop: `text` }] },
        ],
      'SELECT c.id, c.prop FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=5)': [
        { id: `1`, prop: `text`, relatedId: null, embedded: [{ prop: `text`, relatedId: null }] },
        { id: `2`, prop: `text`, relatedId: undefined, embedded: [{ prop: `text`, relatedId: undefined }] },
        { id: `3`, prop: `text`, relatedId: undefined, embedded: [{ prop: `text` }] },
        { id: `4`, prop: `text`, relatedId: undefined, embedded: [{ prop: `text`, relatedId: `5` }] },
        { id: `5`, prop: `text`, embedded: [{ prop: `text` }] },
      ],
    },
    Related: {
      'SELECT c.id, c.prop FROM c WHERE ARRAY_CONTAINS(@p2, c.id) ORDER BY c.id (@p2=5)': [{ id: `5`, prop: `text` }],
    },
  }

  const uc = createUnitTestContext(dummyTypeDefs, responses)

  it(`expects schema to remain the same`, () => {
    const output = printSchemaWithDirectives(uc.schema)
    expect(output).toMatchSnapshot()
  })

  it(`expects meta schema to remain the same`, () => {
    const output = uc.metaSchema
    expect(output).toMatchSnapshot()
  })

  it(`should be retrieve all items`, async () => {
    const query = `
      query {
        dummies {
          page {
            __typename
            id
            prop
            embedded {
              prop
              related {
                __typename
                id
                prop
              }
            }
            related {
              __typename
              id
              prop
              related {
                __typename
                id
                prop
              }
            }
          }
        }
      }
    `
    const result = await uc.execute(query)

    expect(result).toEqual({
      data: {
        dummies: {
          page: [
            {
              __typename: 'Dummy',
              id: `1`,
              prop: `text`,
              embedded: [{ prop: `text`, related: null }],
              related: null,
            },
            {
              __typename: 'Dummy',
              id: `2`,
              prop: `text`,
              embedded: [{ prop: `text`, related: null }],
              related: null,
            },
            {
              __typename: 'Dummy',
              id: `3`,
              prop: `text`,
              embedded: [{ prop: `text`, related: null }],
              related: null,
            },
            {
              __typename: 'Dummy',
              id: `4`,
              prop: `text`,
              embedded: [
                {
                  prop: `text`,
                  related: {
                    __typename: 'Related',
                    id: `5`,
                    prop: `text`,
                  },
                },
              ],
              related: null,
            },
            {
              __typename: 'Dummy',
              id: `5`,
              prop: `text`,
              embedded: [
                {
                  prop: `text`,
                  related: null,
                },
              ],
              related: null,
            },
          ],
        },
      },
    })
  })
})
