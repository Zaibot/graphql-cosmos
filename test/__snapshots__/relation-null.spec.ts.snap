// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nulled relations expects meta schema to remain the same 1`] = `
Object {
  "types": Array [
    Object {
      "container": "Dummies",
      "cosmos": true,
      "database": "Test",
      "external": false,
      "fields": Array [
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "embedded",
          "ours": null,
          "pagination": false,
          "returnMany": true,
          "returnTypename": "Embedded",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "id",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "prop",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "String",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": "Related",
          "cosmos": true,
          "database": "Test",
          "fieldname": "related",
          "ours": "relatedId",
          "pagination": false,
          "returnMany": false,
          "returnTypename": "Related",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Dummy",
    },
    Object {
      "container": null,
      "cosmos": false,
      "database": null,
      "external": false,
      "fields": Array [
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "prop",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "String",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": "Related",
          "cosmos": true,
          "database": "Test",
          "fieldname": "related",
          "ours": "relatedId",
          "pagination": false,
          "returnMany": false,
          "returnTypename": "Related",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Embedded",
    },
    Object {
      "container": null,
      "cosmos": false,
      "database": null,
      "external": false,
      "fields": Array [
        Object {
          "container": "Dummies",
          "cosmos": true,
          "database": "Test",
          "fieldname": "dummies",
          "ours": null,
          "pagination": true,
          "returnMany": true,
          "returnTypename": "Dummy",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Query",
    },
    Object {
      "container": "Related",
      "cosmos": true,
      "database": "Test",
      "external": false,
      "fields": Array [
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "id",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "ID",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": null,
          "cosmos": false,
          "database": null,
          "fieldname": "prop",
          "ours": null,
          "pagination": false,
          "returnMany": false,
          "returnTypename": "String",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
        Object {
          "container": "Related",
          "cosmos": true,
          "database": "Test",
          "fieldname": "related",
          "ours": "relatedId",
          "pagination": false,
          "returnMany": false,
          "returnTypename": "Related",
          "sortOurs": null,
          "sortable": false,
          "theirs": null,
          "whereOps": Array [],
          "whereOurs": null,
        },
      ],
      "filterable": false,
      "sortable": false,
      "typename": "Related",
    },
  ],
}
`;

exports[`Nulled relations expects schema to remain the same 1`] = `
"schema {
  query: Query
}

type Query {
  dummies(cursor: String, limit: Int): DummyPage!
}

type Dummy {
  id: ID!
  prop: String
  embedded: [Embedded]
  related: Related
}

type Embedded {
  prop: String
  related: Related
}

type Related {
  id: ID!
  prop: String
  related: Related
}

type DummyPage {
  total: Int!
  cursor: String
  nextCursor: String
  page: [Dummy]!
}
"
`;
